### PR TITLE
[53] Update compatibility to Sklearn .18.

### PR DESCRIPTION
**Pull request by ajaysaini725**
Currently, spark-sklearn gives deprecation warnings when used with sklearn version .18 because several classes in grid_search and cross_validation were refactored into a new module called model_selection. The changes here make spark-sklearn compatible with the changes introduced in sklearn .18. 

The most critical changes reflected in the new version of sklearn is that:

sklearn.model_selection.GridSearchCV now has:

cv_results_ : dict of numpy (masked) ndarrays - A dict with keys as column headers and values as columns, that can be imported into a pandas DataFrame.
best_estimator_ : estimator - Estimator that was chosen by the search, i.e. estimator which gave highest score (or smallest loss if specified) on the left out data. Not available if refit=False.
best_score_ : float - Score of best_estimator on the left out data.
best_params_ : dict - Parameter setting that gave the best results on the hold out data.
best_index_ : int - The index (of the cv_results_ arrays) which corresponds to the best candidate parameter setting.
scorer_ : function - Scorer function used on the held out data to choose the best parameters for the model.
n_splits_ : int - The number of cross-validation splits (folds/iterations).
While spark-sklearn.GridSearchCV has:

grid_scores_ : list of named tuples
best_estimator_ : estimator - Estimator that was chosen by the search, i.e. estimator which gave highest score (or smallest loss if specified) on the left out data. Not available if refit=False.
best_score_ : float - Score of best_estimator on the left out data.
best_params_ : dict - Parameter setting that gave the best results on the hold out data.
scorer_ : function - Scorer function used on the held out data to choose the best parameters for the model.
The biggest is that sklearn added the more comprehensive cv_results_ which adds data that the formerly compatible grid_scores_ is lacking.

Note: This version of spark-sklearn is not compatible with sklearn <= .17.

### CODE REVIEW

**Comment by ajaysaini725** [comment]
Currently, spark-sklearn gives deprecation warnings when used with sklearn version .18 because several classes in grid_search and cross_validation were refactored into a new module called model_selection. The changes here make spark-sklearn compatible with the changes introduced in sklearn .18. 

The most critical changes reflected in the new version of sklearn is that:

sklearn.model_selection.GridSearchCV now has:

cv_results_ : dict of numpy (masked) ndarrays - A dict with keys as column headers and values as columns, that can be imported into a pandas DataFrame.
best_estimator_ : estimator - Estimator that was chosen by the search, i.e. estimator which gave highest score (or smallest loss if specified) on the left out data. Not available if refit=False.
best_score_ : float - Score of best_estimator on the left out data.
best_params_ : dict - Parameter setting that gave the best results on the hold out data.
best_index_ : int - The index (of the cv_results_ arrays) which corresponds to the best candidate parameter setting.
scorer_ : function - Scorer function used on the held out data to choose the best parameters for the model.
n_splits_ : int - The number of cross-validation splits (folds/iterations).
While spark-sklearn.GridSearchCV has:

grid_scores_ : list of named tuples
best_estimator_ : estimator - Estimator that was chosen by the search, i.e. estimator which gave highest score (or smallest loss if specified) on the left out data. Not available if refit=False.
best_score_ : float - Score of best_estimator on the left out data.
best_params_ : dict - Parameter setting that gave the best results on the hold out data.
scorer_ : function - Scorer function used on the held out data to choose the best parameters for the model.
The biggest is that sklearn added the more comprehensive cv_results_ which adds data that the formerly compatible grid_scores_ is lacking.

Note: This version of spark-sklearn is not compatible with sklearn <= .17.

**Comment by jkbradley** [comment]
CC @dsackin

File: python/spark_sklearn/grid_search.py [hunk 1]
```diff
@@ -42,6 +53,11 @@ class GridSearchCV(BaseSearchCV):
 
     fit_params : dict, optional
         Parameters to pass to the fit method.
+         
+         .. deprecated:: 0.19
```
Comment (jkbradley) [inline python/spark_sklearn/grid_search.py]: This deprecation warning is for a scikit-learn version.  I recommend we just remove the argument from spark-sklearn.

File: python/spark_sklearn/grid_search.py [hunk 2]
```diff
@@ -2,15 +2,26 @@
...
+from sklearn.model_selection import KFold, check_cv, ParameterGrid
+from sklearn.model_selection._validation import _fit_and_score
+from sklearn.model_selection._search import BaseSearchCV, _check_param_grid, _CVScoreTuple
+from sklearn.utils.fixes import MaskedArray
+from sklearn.utils.metaestimators import _safe_split
 from sklearn.utils.validation import _num_samples, indexable
 
+
 class GridSearchCV(BaseSearchCV):
     """Exhaustive search over specified parameter values for an estimator, using Spark to
```
Comment (jkbradley) [inline python/spark_sklearn/grid_search.py]: There are still some differences in this doc when I do a diff between our GridSearchCV and theirs.  This is true for their branches and tags for 0.18, 0.18.X and 0.18.1.  Could you please do a diff and correct the differences wherever it makes sense?

File: python/spark_sklearn/grid_search.py [hunk 3]
```diff
@@ -180,35 +240,40 @@ def fit(self, X, y=None):
         """
         return self._fit(X, y, ParameterGrid(self.param_grid))
 
-    def _fit(self, X, y, parameter_iterable):
-        """Actual fitting,  performing the search over parameters."""
+    
+    def fit(self, X, y=None, groups=None, **fit_params):
```
Comment (jkbradley) [inline python/spark_sklearn/grid_search.py]: Let's keep fit() and _fit() separate, as in sklearn itself.

File: python/spark_sklearn/grid_search.py [hunk 4]
```diff
@@ -180,35 +240,40 @@ def fit(self, X, y=None):
...
+
+      if self.fit_params is not None:
```
Comment (jkbradley) [inline python/spark_sklearn/grid_search.py]: It looks like there are indentation issues here which would make this function do nothing if fit_params is None

File: python/README.md [hunk 5]
```diff
@@ -20,7 +20,7 @@ This package is released under the Apache 2.0 license. See the LICENSE file.
 ## Installation
 
 This package has the following requirements:
- - a recent version of scikit-learn. Version 0.17 has been tested, older versions may work too.
+ - Sklearn version >= 0.18.1
```
Comment (ajaysaini725) [inline python/README.md]: Sklearn 0.18.0 doesn't have MaskedArray which is needed by GridSearchCV so I don't think it would work.

File: python/spark_sklearn/grid_search.py (hunk 3 above)
Comment (ajaysaini725) [inline python/spark_sklearn/grid_search.py]: In sklearn 0.18 they got rid of _fit() and just used fit() (see: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/model_selection/_search.py). I changed it to be consistent with them.

File: python/spark_sklearn/grid_search.py [hunk 6]
```diff
@@ -229,59 +310,93 @@ def fun(tup):
...
+                    rankdata(-array_means, method='min'), dtype=np.int32)
+
+        # Computed the (weighted) mean and std for test scores alone
+        # NOTE test_sample counts (weights) remain the same for all candidates
+        test_sample_counts = np.array(test_sample_counts[:n_splits],
+                                      dtype=np.int)
+
+        _store('test_score', test_scores, splits=True, rank=True,
+               weights=test_sample_counts if self.iid else None)
+        if self.return_train_score:
```
Comment (jkbradley) [inline python/spark_sklearn/grid_search.py]: This differs from 0.18

File: python/spark_sklearn/grid_search.py [hunk 7]
```diff
@@ -229,59 +310,93 @@ def fun(tup):
...
+        _store('fit_time', fit_time)
+        _store('score_time', score_time)
+
+        best_index = np.flatnonzero(results["rank_test_score"] == 1)[0]
+        best_parameters = candidate_params[best_index]
+
+        # Use one MaskedArray and mask all the places where the param is not
+        # applicable for that candidate. Use defaultdict as each candidate may
+        # not contain all the params
+        param_results = defaultdict(partial(MaskedArray,
```
Comment (jkbradley) [inline python/spark_sklearn/grid_search.py]: This differs from 0.18 too

Comment (ajaysaini725) [inline python/spark_sklearn/grid_search.py]: I was working off of version .18.1 because that is the latest published version of sklearn. The .18.1 construction of param_results is different from the construction in .18.0.

File: python/spark_sklearn/grid_search.py (hunk 6 above)
Comment (ajaysaini725) [inline python/spark_sklearn/grid_search.py]: Is this with .18.0 or .18.1?

**Comment by jkbradley** [comment]
Oh!  I just realized: since we're switching from the deprecated GridSearch to the new model_selection.GridSearchCV, we'll have to switch to those tests as well.  test_grid_search_1.py and test_grid_search_2.py do a bit of Python magic to leverage the existing sklearn tests, but hopefully we can just switch to the new sklearn submodule and have it work.
